File: src/registry.ts
import path from "node:path";
import {
  packManifestSchema,
  packRegistrySchema,
  starterRegistrySchema,
  templateManifestSchema,
  type PackManifest,
  type PackRegistryEntry,
  type StarterRegistryEntry,
  type TemplateManifest,
} from "./contracts.js";
import { readJsonFile, repoRoot } from "./fs.js";

async function loadRegistryFile<T>(relativePath: string): Promise<T> {
  return readJsonFile<T>(path.join(repoRoot, relativePath));
}

export async function loadStarterRegistry(): Promise<StarterRegistryEntry[]> {
  const registry = starterRegistrySchema.parse(await loadRegistryFile("registry/starters.json"));
  return registry.starters;
}

export async function loadPackRegistry(): Promise<PackRegistryEntry[]> {
  const registry = packRegistrySchema.parse(await loadRegistryFile("registry/packs.json"));
  return registry.packs;
}

export async function loadStarterManifest(entry: StarterRegistryEntry): Promise<TemplateManifest> {
  return templateManifestSchema.parse(await loadRegistryFile(entry.manifestPath));
}

export async function loadPackManifest(entry: PackRegistryEntry): Promise<PackManifest> {
  return packManifestSchema.parse(await loadRegistryFile(entry.manifestPath));
}

function normalizeRef(ref: string): string {
  const trimmed = ref.trim();
  if (trimmed.startsWith("rendo:")) {
    return trimmed.slice("rendo:".length);
  }

  if (/^https?:\/\//.test(trimmed)) {
    const url = new URL(trimmed);
    const parts = url.pathname.split("/").filter(Boolean);
    return parts[parts.length - 1] ?? trimmed;
  }

  return trimmed;
}

export async function resolveStarterRef(ref: string): Promise<StarterRegistryEntry | null> {
  const registry = await loadStarterRegistry();
  const normalizedRef = normalizeRef(ref);
  return (
    registry.find((item) =>
      [item.id, item.ref, ...item.aliases].some(
        (candidate) => candidate.toLowerCase() === normalizedRef.toLowerCase(),
      ),
    ) ?? null
  );
}

export async function resolvePackRef(ref: string): Promise<PackRegistryEntry | null> {
  const registry = await loadPackRegistry();
  const normalizedRef = normalizeRef(ref);
  return (
    registry.find((item) =>
      [item.name, item.ref, ...item.aliases].some(
        (candidate) => candidate.toLowerCase() === normalizedRef.toLowerCase(),
      ),
    ) ?? null
  );
}

export async function searchRegistry(
  type: "starter" | "pack" | "all",
  keyword: string,
): Promise<Array<Record<string, string | boolean>>> {
  const normalized = keyword.trim().toLowerCase();
  const results: Array<Record<string, string | boolean>> = [];

  if (type === "starter" || type === "all") {
    for (const entry of await loadStarterRegistry()) {
      const manifest = await loadStarterManifest(entry);
      const haystack = [manifest.id, manifest.name, manifest.title, manifest.description].join(" ").toLowerCase();
      if (!normalized || haystack.includes(normalized)) {
        results.push({
          kind: "starter",
          id: manifest.id,
          title: manifest.title,
          version: manifest.version,
          category: manifest.category,
          official: entry.official,
        });
      }
    }
  }

  if (type === "pack" || type === "all") {
    for (const entry of await loadPackRegistry()) {
      const manifest = await loadPackManifest(entry);
      const haystack = [manifest.name, manifest.title, manifest.description].join(" ").toLowerCase();
      if (!normalized || haystack.includes(normalized)) {
        results.push({
          kind: "pack",
          id: manifest.name,
          title: manifest.title,
          version: manifest.version,
          category: manifest.type,
          official: entry.official,
        });
      }
    }
  }

  return results;
}

File: rendo-cli/cli/python/rendo_cli/registry.py
from __future__ import annotations

from .contracts import validate_pack_manifest, validate_template_manifest
from .assets import resolve_asset_path
from .fs import read_json

TEMPLATE_KIND_ALIASES = {
    "starter": "starter-template",
    "starter-template": "starter-template",
    "feature": "feature-template",
    "feature-template": "feature-template",
    "capability": "capability-template",
    "capability-template": "capability-template",
    "provider": "provider-template",
    "provider-template": "provider-template",
    "surface": "surface-template",
    "surface-template": "surface-template",
}


def _load_registry(relative_path: str) -> dict:
    return read_json(resolve_asset_path(relative_path))


def load_template_registry() -> list[dict]:
    return _load_registry("shared/registry/templates.json")["templates"]


def load_pack_registry() -> list[dict]:
    return _load_registry("shared/registry/packs.json")["packs"]


def load_template_manifest(entry: dict) -> dict:
    return validate_template_manifest(read_json(resolve_asset_path(entry["manifestPath"])))


def load_pack_manifest(entry: dict) -> dict:
    return validate_pack_manifest(read_json(resolve_asset_path(entry["manifestPath"])))


def _normalize_ref(ref: str) -> str:
    text = ref.strip()
    if text.startswith("rendo:"):
        return text[len("rendo:") :]
    if text.startswith("http://") or text.startswith("https://"):
        return text.rstrip("/").split("/")[-1]
    return text


def resolve_template_ref(ref: str) -> dict | None:
    normalized = _normalize_ref(ref).lower()
    for item in load_template_registry():
        candidates = [item["id"], item["ref"], *item["aliases"]]
        if any(candidate.lower() == normalized for candidate in candidates):
            return item
    return None


def resolve_starter_ref(ref: str) -> dict | None:
    resolved = resolve_template_ref(ref)
    if resolved is None or resolved["templateKind"] != "starter-template":
        return None
    return resolved


def resolve_template_kind_alias(value: str) -> str | None:
    return TEMPLATE_KIND_ALIASES.get(value.strip().lower())


def resolve_core_template_ref(ref_or_kind: str) -> dict | None:
    resolved = resolve_template_ref(ref_or_kind)
    if resolved is not None:
        return resolved if resolved["templateRole"] == "core" else None

    kind = resolve_template_kind_alias(ref_or_kind)
    if kind is None:
        return None

    for item in load_template_registry():
        if item["templateKind"] == kind and item["templateRole"] == "core":
            return item
    return None


def resolve_pack_ref(ref: str) -> dict | None:
    normalized = _normalize_ref(ref).lower()
    for item in load_pack_registry():
        candidates = [item["name"], item["ref"], *item["aliases"]]
        if any(candidate.lower() == normalized for candidate in candidates):
            return item
    return None


def search_registry(kind: str, keyword: str) -> list[dict]:
    text = keyword.strip().lower()
    results: list[dict] = []
    normalized_kind = kind.lower()

    if normalized_kind != "pack":
        for entry in load_template_registry():
            manifest = load_template_manifest(entry)
            if normalized_kind != "all":
                expected_kind = resolve_template_kind_alias(normalized_kind)
                if expected_kind is None or manifest["templateKind"] != expected_kind:
                    continue
            haystack = " ".join(
                [manifest["id"], manifest["name"], manifest["title"], manifest["description"], *manifest["domainTags"], *manifest["scenarioTags"]]
            ).lower()
            if not text or text in haystack:
                results.append(
                    {
                        "kind": manifest["templateKind"],
                        "id": manifest["id"],
                        "title": manifest["title"],
                        "version": manifest["version"],
                        "category": manifest["category"],
                        "templateKind": manifest["templateKind"],
                        "templateRole": manifest["templateRole"],
                        "official": entry["official"],
                    }
                )

    if normalized_kind in {"pack", "all"}:
        for entry in load_pack_registry():
            manifest = load_pack_manifest(entry)
            haystack = " ".join([manifest["name"], manifest["title"], manifest["description"]]).lower()
            if not text or text in haystack:
                results.append(
                    {
                        "kind": "pack",
                        "id": manifest["name"],
                        "title": manifest["title"],
                        "version": manifest["version"],
                        "category": manifest["type"],
                        "official": entry["official"],
                    }
                )

    return results
